crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4185,6 +4301,89 @@ mod tests {
         Ok(())
     }
 
+    #[test]
+    fn comparison_tuples() -> anyhow::Result<()> {
+        let mut db = setup_db();
+        db.write_dedented(
+            "src/a.py",
+            r#"
+            def int_instance() -> int: ...
+
+            eq1 = (1, (2, "s")) == (1, (2, "s"))
+            eq2 = (1, 2) == (1, 2, 3)
+            eq3 = (1, "s") == (1, 2)
+            eq4 = (int_instance(),) == (1,)
+
+            not_eq1 = (1, (2, "s")) != (1, (2, "s"))
+            not_eq2 = (1, 2) != (1, 2, 3)
+
+            lt1 = (1,) < (1, 2) < (1, 3)
+            lt2 = (1, 2) < (1, 2)
+            lte = (1,) <= (1,) <= (1, 2)
+
+            gt1 = (1, 3) > (1, 2) > (1,)
+            gt2 = (1, 2) > (1, 2)
+            gte = (1, 2) >= (1,) >= (1,)
+            "#,
+        )?;
+
+        assert_public_ty(&db, "src/a.py", "eq1", "Literal[True]");
+        assert_public_ty(&db, "src/a.py", "eq2", "Literal[False]");
+        assert_public_ty(&db, "src/a.py", "eq3", "@Todo"); // TODO: It should be `Unknown`
```

Comment:
This shouldn't be unknown, it should be `Literal[False]` -- we know (or we should know) that `Literal["s"]` and `Literal[2]` cannot be equal.